### DISCUSSION TITLE
authentication process failed

### DISCUSSION THREAD

**Discussion by aryala7** [spatie/laravel-ciphersweet#26] [General]
recently i encrypted my email column in the database and this breaks the authentication process. 
here in the attemptLogin method
`return $this->guard()->attempt(
            $this->credentials($request), $request->filled('remember')
        );`

what i did is to remove the logics and do the job manually. but im not sure if it is a best way to do that. 
`$operator = Operator::query()->whereBlind('email','operator_email_index',$request['email'])->first();
        if($operator && Hash::check($request['password'],$operator->password)){
            return true;
        } 
        return false;
`

could you please help me with this scenario?

**Comment by aryala7** [comment]
i somehow handled this in a dirty way.
before using laravel fortify i have to first get the user with its email using raw query to prevent triggering the model observer:
`$operator = Operator::query()->whereBlind('email','operator_email_index', strtolower($request['email']))->first();`
if operator exists then overwrite the email request and get the email directly from model:
`if($operator) {
            $credentials['email'] = DB::table('operator_operators')->where('id',$operator->id)->first()->email;
        }`
        and then use the default Auth::Attempt method
        `$this->guard()->attempt(
            $credentials, $request->filled('remember')
        );`

but with **fortify** you can easily overwrite the attempt with _**authenticateUsing**_ method and prevent the extra queries.

**Comment by gawsoftpl** [comment]
Add to user model static method findByEmail. This method will search user by email using blindIndex
```php
   public static function findByEmail(string $email)
    {
        return User::whereBlind('email', 'email_index', $email)
            ->first();
    }
```

In Auth attempt you can use Closure

```php
Auth::attempt(
            [
                function(Builder $query){
                    $user = User::findByEmail($this->email);
                    if (!$user) throw ValidationException::withMessages(['email' => 'Invalid email or password']);
                    $query->where('id', $user->id);
                    return $query;
                },
                'password' => $this->password, //auto hashed password in User model,
            ], $this->remember)) {
```
